#36 Make Content-Type determination case-insensitive

ljcoomber opened the pull request:
Fix for #36.

Verified by modifying existing test - happy to add an explicit test instead if you suggest where it should go.

ceilfors commented:
Thanks for the PR @ljcoomber! Looking through the code, I'm quite interested to fix the case insensitivity once and for all, hence a few things:

1. If headers are coming through as `content-type`, users should be able to retrieve the header with `Content-Type` (This is the bug)
2. If headers are coming through as `Content-Type`, users should be able to retrieve the header with `content-type` (This is currently implemented)

The PR at the moment is only fixing the framework, not for the users at the moment. To fix this both for the framework and users:
1. There is already a class `ApiGatewayInputHeaders` at the moment that is responsible to help on case sensitivity for the users. We can override the `event.headers` in `ApiGatewayEvent.fromRaw` so that this is used by the framework too.
2. To fix the bug `ApiGatewayInputHeaders` must be able to retrieve a property with case insensitivity

To retrieve the property case insensitively, a suggestion, you can return a Proxy object inside the `constructor` of `ApiGatewayInputHeaders`, like this (not tested):
```
const handler = {
  get: function(target, prop) {
    return target[prop.toLowerCase()];
  }
};

return new Proxy(this, handler);
```

Beware that the above may break the user land if they are accessing original JavaScript object property, so you will have to detect if the `prop` is part of the original header i.e. by using `Object.prototype.hasOwnProperty()` or something.

On the test, you should be able to add unit tests under:
1. `ApiGatewayInputHeaders` for the new functionality
2. `ApiGatewayEvent` to ensure that parseRequestBody is able to fix the original issue

I hope this helps, happy to support the PR further.

ljcoomber commented:
Hi, I've updated the PR based on your suggestions.

I wanted to share the instance of ApiGatewayInputHeaders to avoid creating needless duplication. This meant altering the event construction order and passing the ApiGatewayInputHeaders into the parse function.

As part of this I dropped the explicit event constructor. I would normally assume a constructor contains everything an object needs to be valid, but was not so in this case. Using the implicit no-args constructor therefore seems fine.

> Beware that the above may break the user land if they are accessing original JavaScript object property, so you will have to detect if the prop is part of the original header i.e. by using Object.prototype.hasOwnProperty() or something.

I didn't understand this comment - please can you explain?

ceilfors commented on packages/laconia-event/src/apigateway/ApiGatewayInputHeaders.js:
```diff
@@ -4,5 +4,13 @@ module.exports = class ApiGatewayInputHeaders {
     Object.entries(eventHeaders).forEach(([headerName, headerValue]) => {
```
We shouldn't need the line above anymore i.e. `Object.assign(this, eventHeaders);`

ceilfors commented on packages/laconia-event/src/apigateway/parseRequestBody.js:
```diff
@@ -13,8 +13,8 @@ const parseJsonBody = body => {
 
-module.exports = event => {
-  const contentType = event.headers["Content-Type"] || "";
+module.exports = (event, headers) => {
```
It's a good change! This is internal function anyway, it shouldn't break any backward compatibility

ceilfors commented on packages/laconia-event/test/apigateway/parseRequestBody.spec.js:
```diff
@@ -68,8 +76,23 @@ describe("parseRequestBody", () => {
     });
   });
+
+  describe("when content type header is lowercase", () => {
```
Great test!

ceilfors commented:
I see that fossa is failing the build, so I have now removed it from our CI pull request build. If you take the latest commit from master, your build should pass.

ljcoomber commented:
Made the suggested changes, and it was a good catch - it would have been very confusing for the user if Object properties / methods stopped working.

I think the Git history problem was me - sorry! I originally rebased your later changes from master but the conflicts meant it failed. I thought I had successfully rolled back and merged  but I may have screwed that up.

CircleCI still failing, but it looks like a problem running the coverage tests which succeed when run locally (using `npm test`). From the error, it looks like a problem connecting to some dependency?

ceilfors commented on packages/laconia-event/src/apigateway/ApiGatewayInputHeaders.js:
```diff
@@ -1,8 +1,18 @@
+      get: (target, prop) => {
+        const isSymbol = typeof prop === "symbol";
+        const isExisting = prop in target;
+        const propertyKey = isSymbol || isExisting ? prop : prop.toLowerCase();
```
What's the benefit of checking it like this, instead of doing `Object.prototype.hasOwnProperty.call(target, prop) ? prop.toLowerCase() : prop` ?

ljcoomber replied:
Yes, happy for you to proceed.

The reason for the check above is that `hasOwnProperty()` does not check for inherited properties, hence things like `toString()` will stop working because the Proxy will change it to `tostring()` (from my rudimentary understanding of JS)

ceilfors commented:
@all-contributors please add @ljcoomber for code
